sherpa/astro/ui/utils.py

```diff
@@ -9261,27 +9275,92 @@ def fake_pha(self, id, arf, rmf, exposure, backscal=None, areascal=None,
-        bkg_models = {}
-        if bkg is not None:
-            if bkg == 'model':
-                bkg_models = {1: self.get_bkg_source(id)}
-            else:
-                for bkg_id in d.background_ids:
-                    d.delete_background(bkg_id)
-                self.set_bkg(id, bkg)
+        # If bkg is None then there is no background component, which
+        # means removing any background dataset or models.  They will
+        # need to be added back after the call to fake_pha.
+        #
+        # If bkg="model" then this is already included (as long as
+        # background models are set, but we do not force this
+        # requirement here).
+        #
+        # If bkg is a DataPHA object then remove all the existing
+        # backgrounds and replace them with the background, and set
+        # include_bkg_data. There is also the need to restore a
+        # background model if set: given that there's only the
+        # possibility of using a single background dataset this
+        # potentially loses information, but it's unclear what the
+        # user really expects here given the existing API.
+        #
+        include_bkg_data = False
+        restore = {}
+        old_model = None
+        if bkg is None:
+            # Remove the background components to try to avoid
+            # potential confusion with downstream processing. There is
+            # an argument to say they should be kept, but it's not
+            # clear what is best.
+            #
+            for bkg_id in d.background_ids:
+                restore[bkg_id] = {"data": d.get_background(bkg_id)}
+                try:
+                    restore[bkg_id]["model"] = self.get_bkg_source(idval, bkg_id)
+                    self.delete_bkg_model(idval, bkg_id)
+                except ModelErr:
+                    pass
 
-        # Calculate the source model, and take a Poisson draw based on
-        # the source model.  That becomes the simulated data.
-        m = self.get_model(id)
+                d.delete_background(bkg_id)
 
-        fake.fake_pha(d, m, is_source=False, add_bkgs=bkg is not None,
-                      id=str(id), bkg_models=bkg_models, method=method)
+        elif bkg != "model":
+            try:
+                # Note: this fails if set_bkg_full_model was used (in
+                # that the old model expression will be lost).
+                #
+                old_model = self.get_bkg_source(idval)
+            except ModelErr as me:
+                old_model = None
+
+            for bkg_id in d.background_ids:
+                self.delete_bkg_model(idval, bkg_id)
+                d.delete_background(bkg_id)
+
+            self.set_bkg(idval, bkg)
+            include_bkg_data = True
+
+        # The source model (which includes any background model
+        # components and the response) is used. The decision here is
+        # whether to set the include_bgnd_flag to True or not. The
```

Comment:
Do you mean
```suggestion
        # whether to set the include_bkg_data to True or not. The
```

Reply:
During development I switched flag names, so I think this is an incorrectly-updated comment. Or I always had it not documenting reality.

Reply:
Hmmmm. I need to investigate further. I have in line 9294 `include_bkg_data` set False, then in 9327 it can be set True, but then in line 9340 it's unconditionally set to False, which means the earlier code is pointless. 

So either this extra text and setting is wrong and can be removed, or the earlier stuff is wrong. Or they are noth essentially the same (but you cna see they are not quite the same).

Reply:
Hmm. So I have added tests that check this, and they show that both approaches work. I need to just look into why this is, as from a brief look on Sunday it wasn't obvious that they would.

Reply:
Doh. They are the same logic.

Reply:
I've removed the second `include_bkg_data` set of changes as they are unnecessary - hence the comment in question has been completely removed. There's also a little-bit if comment clean up (and some code-style changes to use three rather than one-letter variable names (d/m vs pha/mdl).